env/env_encrypt2.cc

```diff
@@ -0,0 +1,572 @@
+  //
+  // AES_BLOCK_SIZE assumed to be 16
+  //
+  assert(AES_BLOCK_SIZE == 16);
+  assert(sizeof(AesAlignedBlock) == AES_BLOCK_SIZE);
+
+  Status status;
+  ALIGN16 AesAlignedBlock block_in, block_out, iv;
+  int out_len = 0, in_len = {AES_BLOCK_SIZE}, ret_val;
+
+  if (EncryptedEnv2::crypto_.IsValid()) {
+    // make a context once per thread
+    if (!aes_context) {
+      aes_context = std::unique_ptr<EVP_CIPHER_CTX, void (*)(EVP_CIPHER_CTX*)>(
+          EncryptedEnv2::crypto_.EVP_CIPHER_CTX_new(),
+          EncryptedEnv2::crypto_.EVP_CIPHER_CTX_free_ptr());
+    }
+
+    ret_val = EncryptedEnv2::crypto_.EVP_CIPHER_CTX_reset(aes_context.get());
+    if (1 == ret_val) {
+      // https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38a.pdf
+      memcpy(iv.bytes, nonce_, AES_BLOCK_SIZE / 2);
+      EncodeFixed64((char*)&iv.bytes[AES_BLOCK_SIZE / 2],
+                    blockIndex);  // this will be little endian
+      block_in.nums[0] = 0;
+      block_in.nums[1] = 0;
+
+      ret_val = EncryptedEnv2::crypto_.EVP_EncryptInit_ex(
+          aes_context.get(), EncryptedEnv2::crypto_.EVP_aes_256_ctr(), nullptr,
+          key_.key, iv.bytes);
+      if (1 == ret_val) {
+        ret_val = EncryptedEnv2::crypto_.EVP_EncryptUpdate(
+            aes_context.get(), block_out.bytes, &out_len, block_in.bytes,
+            in_len);
+
+        if (1 != ret_val || AES_BLOCK_SIZE != out_len) {
+          status = Status::InvalidArgument("EVP_EncryptUpdate failed: ",
+                                           AES_BLOCK_SIZE == out_len
+                                               ? "bad return value"
+                                               : "output length short");
+        }
+      } else {
+        status = Status::InvalidArgument("EVP_EncryptInit_ex failed.");
+      }
+    } else {
+      status = Status::InvalidArgument("EVP_CIPHER_CTX_reset failed.");
+    }
+
+    // XOR data with ciphertext.
+    uint64_t* data_ptr;
+    data_ptr = (uint64_t*)data;
+    *data_ptr ^= block_out.nums[0];
+    data_ptr = (uint64_t*)(data + 8);
+    *data_ptr ^= block_out.nums[1];
+  } else {
+    status = Status::NotSupported(
+        "libcrypto not available for encryption/decryption.");
+  }
+
+  return status;
+}
+
+Status AESBlockAccessCipherStream::DecryptBlock(uint64_t blockIndex, char* data,
+                                                char* scratch) {
+  return EncryptBlock(blockIndex, data, scratch);
+}
+
+Status CTREncryptionProvider2::CreateNewPrefix(const std::string& /*fname*/,
+                                               char* prefix,
+                                               size_t prefixLength) {
+  Status s;
+  if (EncryptedEnv2::crypto_.IsValid()) {
+    if (sizeof(Prefix0) <= prefixLength) {
+      int ret_val;
+
+      Prefix0* pf = {(Prefix0*)prefix};
+      memcpy(pf->key_description_, key_desc_.desc, sizeof(key_desc_.desc));
+      ret_val = EncryptedEnv2::crypto_.RAND_bytes(
```

Comment:
If we are going to fix item 1, shall we generate a full 16 bytes initial IV?

Reply:
I have zero hesitations in matching openssl equivalent.  Any chance you can point to openssl source code so that I can copy and therefore reduce the number of cycles here?

Reply:
This is not to match openssl, as IVs are provided by caller. But since IV (used as both of nonce and counter) is just a big 128-bit integer, I feel it more natural to rand-gen full 16 bytes IV.

Reply:
The 128-bit IV thing you can refer to this: https://github.com/openssl/openssl/blob/master/crypto/modes/ctr128.c#L26

Reply:
This is done.  However, the new support routine  BigEndianAdd128() has a carry problem beyond the lower 8 bytes.  I am addressing this bug plus adding unit tests for this routine.